Implement a functioning Blood Tank

Arcaratus opened the pull request:
-Added a search bar to the Upgrade Tomes Creative Tab
-Updated some Altar fluid code (remove deprecated stuff)
-Moved Rendering classes into appropriate package
-Fix the localization errors on the Demon Crystals
-A few cleanups here and there
-Removed obsolete fluid methods in Utils
-Has comparator output
-Update JEI compat

-- TO DO --
-Add recipes for the Blood Tank
~~-Fix the lighting-not-updating-upon-fluid-placement within the tank bug~~
~~-Render tank in inventory if possible~~ <- Terrible but working
-Future (maybe 1.11) update from deprecated methods and classes throughout entire mod
~~-Fix JEI bug where cheating a tank in reverts to a Tier 1 tank regardless of which tank you cheated in~~

Arcaratus committed:
Implement a functioning Blood Tank
Added a search bar to the Upgrade Tomes Creative Tab
Updated some Altar fluid code (remove deprecated stuff)
Moved Rendering classes into appropriate package
Fix the localization errors on the Demon Crystals
A few cleanups here and there

Arcaratus committed:
Did the requested changes

Arcaratus committed:
Merge branch '1.9' of https://github.com/WayofTime/BloodMagic into 1.9

TehNut commented:
Again, please no TESR. Use the model system.

I will look over this more when I get home

TehNut commented on src/main/java/WayofTime/bloodmagic/compat/waila/provider/DataProviderBloodTank.java:
```diff
@@ -0,0 +1,78 @@
+        if (!config.getConfig(Constants.Compat.WAILA_CONFIG_BLOOD_TANK))
+            return currenttip;
+
+        if (accessor.getPlayer().isSneaking() || config.getConfig(Constants.Compat.WAILA_CONFIG_BYPASS_SNEAK))
```
This whole bit is wrong. You cannot access a Tiles NBT here. Write everything you need to the tag in `#getNBTData` and read it from `accessor#getNBTData`. I need to fix our other implementations to do this. See the altar impl for an example.

TehNut commented on src/main/java/WayofTime/bloodmagic/item/block/ItemBlockBloodTank.java:
```diff
@@ -17,108 +25,144 @@
         super(block);
+
+        setHasSubtypes(true);
+        setMaxDamage(0);
```
Unecessary.

TehNut commented on src/main/resources/assets/bloodmagic/lang/en_US.lang:
```diff
@@ -254,11 +254,11 @@ tile.BloodMagic.path.obsidiantile.name=Tiled Obsidian Path
+tile.BloodMagic.demonCrystal.destructive.name=Destructive Will Crystal Cluster
+tile.BloodMagic.demonCrystal.vengeful.name=Vengeful Will Crystal Cluster
+tile.BloodMagic.demonCrystal.steadfast.name=Steadfast Will Crystal Cluster
 
```
If you change this here, you need to change every language.

Arcaratus committed:
Merge remote-tracking branch 'origin/1.9' into 1.9

Arcaratus committed:
Add alpha rendering

WayofTime commented:
Good to pull, or need more stuff?

TehNut commented:
Needs tank rewritten as a model instead of a TESR. If this is being done, it should be done correctly.

WayofTime commented:
@TehNut Talked to both Grum and cpw. They said that a TESR for any form of block that updates its model regularly is a bad idea because it redraws the chunk of 16^3 blocks that it is in. So the static part of the model (the block) should be a model, but the fluid part should either be a regular TESR or a fast TESR.

TehNut commented:
Alright, let's do that, then.

Arcaratus committed:
Do the fixes

Arcaratus committed:
Merge remote-tracking branch 'origin/1.9' into 1.9

TehNut commented on src/main/resources/assets/bloodmagic/blockstates/BlockBloodTank.json:
```diff
@@ -0,0 +1,92 @@
+    "tier": {
+	  "0": {
+		"textures": {
+	      "all": "bloodmagic:blocks/BloodTank"
```
If you add 

``` json
"textures": {
    "all": "bloodmagic:blocks/BloodTank"
}
```

to the `"default": {}` block, you shouldn't need it in all of these variants.

TehNut commented on src/main/java/WayofTime/bloodmagic/compat/waila/provider/DataProviderBloodTank.java:
```diff
@@ -0,0 +1,78 @@
+                currenttip.add(TextHelper.localizeEffect("tooltip.BloodMagic.fluid.capacity") + ": " + bloodTank.getCapacity() + "mB");
+
+                NBTTagCompound tag = accessor.getNBTData().getCompoundTag(Constants.NBT.TANK);
+                if (!Strings.isNullOrEmpty(tag.getString("FluidName")) && FluidStack.loadFluidStackFromNBT(tag) != null)
```
Save the `FluidStack` instead of creating it from NBT twice.

TehNut commented on src/main/java/WayofTime/bloodmagic/compat/waila/provider/DataProviderBloodTank.java:
```diff
@@ -0,0 +1,78 @@
+                if (!Strings.isNullOrEmpty(tag.getString("FluidName")) && FluidStack.loadFluidStackFromNBT(tag) != null)
+                {
+                    currenttip.add(TextHelper.localizeEffect("tooltip.BloodMagic.fluid.type") + ": " + FluidStack.loadFluidStackFromNBT(tag).getLocalizedName());
+                    currenttip.add(TextHelper.localizeEffect("tooltip.BloodMagic.fluid.amount") + ": " + tag.getInteger("Amount") + "/" + bloodTank.getCapacity() + "mB");
```
This `getCapacity()` call (and the one above) need to either be changed to be retrieved from NBT or the `getCapacity()` method needs to be changed to not depend on a field obtained via NBT.

TehNut commented on src/main/java/WayofTime/bloodmagic/client/render/block/RenderBloodTank.java:
```diff
@@ -0,0 +1,131 @@
+        tessellator.draw();
+    }
+
+    public static TextureAtlasSprite checkIcon(TextureAtlasSprite icon)
```
Why create this method? You run this bit of code once, as far as I can tell.

TehNut commented:
Whitespacing seems to be a mix of tabs and spaces. Please convert all of the tabs to spaces.

WayofTime commented:
That said, I'll probably be running Eclipse's formatter afterwards, so the last part isn't as bad of an issue.

Arcaratus committed:
Merge remote-tracking branch 'origin/1.9' into 1.9

Arcaratus committed:
Update JEI compat
Fixed JEI cheating in bug
Updated fluid handling
Removed unnecessary fluid methods in Utils

Arcaratus committed:
Added comparator output to the Blood Tank
Fixed the lighting bugs

Arcaratus committed:
Implement item-rendering for the Blood Tank

Arcaratus committed:
Remove InventoryRenderHelper usage

Arcaratus committed:
Merge branch '1.9' of https://github.com/WayofTime/BloodMagic into 1.9

TehNut commented on src/main/java/WayofTime/bloodmagic/compat/jei/BloodMagicPlugin.java:
```diff
@@ -99,4 +99,9 @@ public void register(@Nonnull IModRegistry registry)
     }
+
+    @Override
+    public void onRuntimeAvailable(IJeiRuntime runtime)
```
Why implement this? `BlankModPlugin` should be taking care of any unneeded methods.

TehNut commented on src/main/java/WayofTime/bloodmagic/util/handler/event/BakedCustomItemModel.java:
```diff
@@ -0,0 +1,205 @@
+
+/**
+ * Shamelessly taken off of the Mekanism repository written by aidancbrady
+ * https://github.com/aidancbrady/Mekanism
```
`@author aidancbrady`

Arcaratus committed:
Merge remote-tracking branch 'origin/1.9' into 1.9

Arcaratus committed:
Moving files...

Arcaratus committed:
Merge remote-tracking branch 'origin/1.9' into 1.9
